### PR TITLE
Delete Access Token

### LINKED ISSUE #941: method to delete AccessTokens on document.Model

### PR DESCRIPTION
**Pull request by charleenfei**
<This is intended as a guide when describing-reviewing pull requests for go-centrifuge>
Closes #941 .

### CODE REVIEW

File: documents/read_acls.go [hunk 1]
```diff
@@ -452,6 +452,43 @@ func (cd *CoreDocument) AddAccessToken(ctx context.Context, payload documentpb.A
...
+	}
+
+	at := ncd.Document.AccessTokens
+	found := false
+	for i, t := range at {
+		if hexutil.Encode(t.Grantee) == granteeID {
+			new := removeTokenAtIndex(i, at)
+			ncd.Document.AccessTokens = new
+			ncd.Modified = true
+			found = true
```
Comment (manuelpolzhofer) [inline documents/read_acls.go]:
```suggestion
			return ncd, nil
```

File: documents/read_acls.go [hunk 2]
```diff
@@ -452,6 +452,43 @@ func (cd *CoreDocument) AddAccessToken(ctx context.Context, payload documentpb.A
...
+		}
+	}
+	if !found {
```
Comment (manuelpolzhofer) [inline documents/read_acls.go]: you can remove the rest and just `return nil, ErrAccessTokenNotFound`

File: documents/read_acls.go [hunk 3]
```diff
@@ -452,6 +452,43 @@ func (cd *CoreDocument) AddAccessToken(ctx context.Context, payload documentpb.A
...
+	if err != nil {
+		return nil, err
+	}
+
+	_, err = identity.StringsToDIDs(granteeID)
+	if err != nil {
+		return nil, err
+	}
+
+	at := ncd.Document.AccessTokens
```
Comment (manuelpolzhofer) [inline documents/read_acls.go]: why are you calling it `at`. It is a list of tokens.

maybe `accessTokens`

File: documents/read_acls.go [hunk 4]
```diff
@@ -452,6 +452,43 @@ func (cd *CoreDocument) AddAccessToken(ctx context.Context, payload documentpb.A
...
+	}
+	if !found {
+		return nil, ErrAccessTokenNotFound
+	}
+	return ncd, nil
+}
+
+// RemoveTokenAtIndex removes the access token at index i from slice a
+// Note: changes the order of the slice elements
+func removeTokenAtIndex(i int, tokens []*coredocumentpb.AccessToken) []*coredocumentpb.AccessToken {
```
Comment (manuelpolzhofer) [inline documents/read_acls.go]:
```suggestion
func removeTokenAtIndex(idx int, tokens []*coredocumentpb.AccessToken) []*coredocumentpb.AccessToken {
```

I only would use `i` inside a loop

File: documents/read_acls.go [hunk 5]
```diff
@@ -452,6 +452,43 @@ func (cd *CoreDocument) AddAccessToken(ctx context.Context, payload documentpb.A
...
+	tokens[i] = tokens[len(tokens)-1]
```
Comment (manuelpolzhofer) [inline documents/read_acls.go]: shorter version :)
```golang
tokens[len(s)-1], tokens[idx] = tokens[idx], tokens[len(s)-1]
return tokens[:len(s)-1]
```

File: documents/read_acls.go [hunk 6]
```diff
@@ -452,6 +452,43 @@ func (cd *CoreDocument) AddAccessToken(ctx context.Context, payload documentpb.A
...
+	tokens[len(tokens)-1] = nil
+	tokens = tokens[:len(tokens)-1]
+	return tokens
```
Comment (vimukthi-git) [inline documents/read_acls.go]: https://github.com/golang/go/wiki/SliceTricks

File: documents/read_acls.go [hunk 7]
```diff
@@ -452,6 +452,37 @@ func (cd *CoreDocument) AddAccessToken(ctx context.Context, payload documentpb.A
...
+
+	_, err = identity.StringsToDIDs(granteeID)
+	if err != nil {
+		return nil, err
+	}
+
+	accessTokens := ncd.Document.AccessTokens
+	for i, t := range accessTokens {
+		if hexutil.Encode(t.Grantee) == granteeID {
+			new := removeTokenAtIndex(i, accessTokens)
```
Comment (manuelpolzhofer) [inline documents/read_acls.go]:
```suggestion
			ncd.Document.AccessTokens := removeTokenAtIndex(i, accessTokens)
```

File: documents/read_acls.go [hunk 8]
```diff
@@ -452,6 +452,37 @@ func (cd *CoreDocument) AddAccessToken(ctx context.Context, payload documentpb.A
...
+			ncd.Document.AccessTokens = new
```
Comment (manuelpolzhofer) [inline documents/read_acls.go]:
```suggestion

```

File: documents/read_acls.go [hunk 9]
```diff
@@ -452,6 +452,37 @@ func (cd *CoreDocument) AddAccessToken(ctx context.Context, payload documentpb.A
...
+		return nil, err
+	}
+
+	_, err = identity.StringsToDIDs(granteeID)
+	if err != nil {
+		return nil, err
+	}
+
+	accessTokens := ncd.Document.AccessTokens
+	for i, t := range accessTokens {
```
Comment (manuelpolzhofer) [inline documents/read_acls.go]:
```suggestion
	for _, t := range accessTokens {
```

**Pull request merged**